### PR TITLE
production pipeline files

### PR DESCRIPTION
**Pull request by AliceCMos**

### CODE REVIEW

File: costa_pipeline_production/criteria.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    total_days_termination=21           : int   # max days in block
+    """
+
+
+@schema
+class ScaraCriteria(dj.Imported):
+    definition = """
+    -> acquisition.Session.Scara
+    ---
+    """
```
Comment (shenshan) [inline costa_pipeline_production/criteria.py:45]: The make function of this table heavily rely on `ScaraCriteriaParameters` and `scara.ScaraPerformance`. I would suggest to put some dependencies in the table definition. If `scara.ScaraPerformance` is not populated, `ScaraCriteria` should not be able to populate.

**Review by shenshan — CHANGES_REQUESTED** [review]
Great work folks. I left some comments on the design consideration of a few tables, common patterns of dj and python usage and some minor details.

File: costa_pipeline_production/criteria.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
```
Comment (shenshan) [inline costa_pipeline_production/criteria.py:46]: key_source = scara.ScaraPerformance & ScaraCriteriaParameters
----
-> ScaraCriteriaParameters
-> scara.ScaraPerformance

File: costa_pipeline_production/criteria.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    class ScaraCriteriaPhase3(dj.Part):
+        definition = """
+        -> master
+        ---
+        #-> acquisition.Session.proj(session_number='session_number')
+        three_day_average=null    : float
+        start_new_target          : boolean
+        """
+
+    key_source = acquisition.Session & acquisition.Session.Scara
```
Comment (shenshan) [inline costa_pipeline_production/criteria.py:62]: could be deleted

File: costa_pipeline_production/instrument.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        ---
+        -> hardware.Filter
+        uninstall_date=null : date
+        """
+
+# ----------------------------------------------------------
+@schema
+class Optogenetics(dj.Lookup):
+    definition = """
+    opto_id:   varchar(64) # e.g. Driver number?
```
Comment (shenshan) [inline costa_pipeline_production/instrument.py:49]: rename to something like `opto_setup_name`

File: costa_pipeline_production/lab.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        ['SC2.081',	'Mouse housing'],
+        ['SC4.100',	'Mouse housing'],
+        ['SC4.050',	'Behavioral studies, imaging 2Photon'],
+        ['SC4.103',	'Behavioral studies, imaging, euthanasia'],
+        ['SC4.103A','Euthanasia'],
+        ['SC4.104',	'Behavioral studies, imaging, euthanasia'],
+        ['SC4.106',	'Behavioral studies, imaging, euthanasia'],
+        ['SC4.107',	'Behavioral studies, imaging, euthanasia']
+        ]
+
```
Comment (shenshan) [inline costa_pipeline_production/lab.py:113]: remove the room numbers

File: costa_pipeline_production/opto_stimulus_scara.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    opto_duration:      int  # (msec) total pulse train duration
+    opto_frequency:     int  # (Hz)
+    opto_pulsewidth:    int  # (msec) of LED on per pulse
+    opto_train_rampup_duration=null     : int
+    opto_train_rampdown_duration=null   : int
+    opto_pulse_rampup_duration=null     : int
+    opto_pulse_rampdown_duration=null   : int
+    """
+
+key_source = acquisition.Session & acquisition.Session.Optogenetics
```
Comment (shenshan) [inline costa_pipeline_production/opto_stimulus_scara.py:21]: could be removed

File: costa_pipeline_production/scara.py [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+        # ------------ Read Task File --------------
+        phase = (acquisition.Session.Scara & key).fetch1('phase')
+
+        if phase == 'Phase 1':
+            data = pd.read_csv(task_file_fullpath, sep=",", nrows=13, error_bad_lines=False, warn_bad_lines=False)
+        elif phase == 'Phase 2':
+            data = pd.read_csv(task_file_fullpath, sep=",", nrows=18, error_bad_lines=False, warn_bad_lines=False)
+        elif phase == 'Phase 3':
+            data = pd.read_csv(task_file_fullpath, sep=",", nrows=19, error_bad_lines=False, warn_bad_lines=False)
+
```
Comment (shenshan) [inline costa_pipeline_production/scara.py:183]:
```
phase_row = {
'Phase 1' : 13,
'Phase 2' : 18,
'Phase 3' : 19
}
data = pd.read_csv(task_file_fullpath, sep=",", nrows=phase_row[phase], error_bad_lines=False, warn_bad_lines=False)
```

File: costa_pipeline_production/suite2p.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            }
+
+        language_extension = 'ijm'
+        result_script = ij.py.run_script(language_extension,macros_raw2ometiff, args)
+        print(result_script.getOutput('message'))
+
+        # ------------ Insert into table --------------
+        name = (acquisition.Session.TwoPhoton & key).fetch1('two_photon_name')
+
+        if len(result_script) > 1: #only populate if it ran
```
[the remark this answers is not in the document]
Comment (shenshan) [reply costa_pipeline_production/suite2p.py:124]: `if len(result_script):`

File: costa_pipeline_production/suite2p.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    ---
+    ops_filename: varchar(200)
+   """
+
+@schema
+class Suite2pRun(dj.Imported):
+    definition = """
+    -> acquisition.Session.TwoPhoton
+    ---
+    run_date=CURRENT_TIMESTAMP:  datetime
```
[the remark this answers is not in the document]
Comment (shenshan) [reply costa_pipeline_production/suite2p.py:147]: -> Suite2pParameters

File: costa_pipeline_production/twophoton.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    """
+    key_source = acquisition.Session & acquisition.Session.TwoPhoton
+    def make(self, key):
+        microscope = (acquisition.Session.TwoPhoton & key).fetch1('two_photon_name')
+        if microscope == 'ThorlabsBergamo':
+           #Construct filenames:
+            session_date = (acquisition.Session & key).fetch1('session_start_time')
+            subject_id = (acquisition.Session & key).fetch1('subject_id')
+            #Raw File Name eg: D1-Cre1234_2020-06-23_0001.avi
+            raw_file_name = str(subject_id) + '_' + str(session_date.date()) + '_0001' +'.raw'
```
[the remark this answers is not in the document]
Comment (shenshan) [reply costa_pipeline_production/twophoton.py:40]: ''.format()

File: costa_pipeline_production/twophoton.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+            fullname = user_name.split()  #split into first and last name
+            firstname = fullname[0]     #get firstname
+
+            #Get project name:
+            experiment_name = (subject.Subject & key).fetch1('experiment_nickname')
+            cohort_number = (subject.Subject & key).fetch1('cohort_number')
+            project_name = experiment_name + '-'+ str(cohort_number)
+
+            current_os = platform.system()
+            if current_os == "Darwin":
```
[the remark this answers is not in the document]
Comment (AliceCMos) [reply costa_pipeline_production/twophoton.py:60]: yes, at the moment only scara experiments are run on this microscope. But will make it more general

**Pull request merged**